soc/arm/nxp_imx/rt/power_rt10xx.c

```diff
@@ -0,0 +1,260 @@
+	irq_unlock(0);
+	__DSB();
+	__ISB();
+
+	if (mode == kCLOCK_ModeWait) {
+		/* Clear the SLEEPDEEP bit to go into sleep mode (WAIT) */
+		SCB->SCR &= ~SCB_SCR_SLEEPDEEP_Msk;
+	} else {
+		/* Set the SLEEPDEEP bit to enable deep sleep mode (STOP) */
+		SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;
+	}
+	/* WFI instruction will start entry into WAIT/STOP mode */
+	__WFI();
+
+}
+
+static void lpm_set_run_mode_config(void)
+{
+	/* Clear GPC wakeup source */
+	GPC_DisableIRQ(GPC, DT_IRQN(DT_INST(0, nxp_gpt_hw_timer)));
+	CCM->CLPCR &= ~(CCM_CLPCR_LPM_MASK | CCM_CLPCR_ARM_CLK_DIS_ON_LPM_MASK);
+}
+
+/* Toggle the analog bandgap reference circuitry on and off */
+static void bandgap_set(bool on)
+{
+	if (on) {
+		/* Enable bandgap in PMU */
+		PMU->MISC0_CLR = PMU_MISC0_REFTOP_PWD_MASK;
+		/* Wait for it to stabilize */
+		while ((PMU->MISC0 & PMU_MISC0_REFTOP_VBGUP_MASK) == 0) {
+
+		}
+		/* Disable low power bandgap */
+		XTALOSC24M->LOWPWR_CTRL_CLR = XTALOSC24M_LOWPWR_CTRL_LPBG_SEL_MASK;
+	} else {
+		/* Disable bandgap in PMU and switch to low power one */
+		XTALOSC24M->LOWPWR_CTRL_SET = XTALOSC24M_LOWPWR_CTRL_LPBG_SEL_MASK;
+		PMU->MISC0_SET = PMU_MISC0_REFTOP_PWD_MASK;
+	}
+}
+
+/* Should only be used if core clocks have been reduced- drops SOC voltage */
+static void lpm_drop_voltage(void)
+{
+	/* Move to the internal RC oscillator, since we are using low power clocks */
+	CLOCK_InitRcOsc24M();
+	/* Switch to internal RC oscillator */
+	CLOCK_SwitchOsc(kCLOCK_RcOsc);
+	CLOCK_DeinitExternalClk();
+	/*
+	 * Change to 1.075V SOC voltage. If you are experiencing issues with
+	 * low power mode stability, try raising this voltage value.
+	 */
+	DCDC_AdjustRunTargetVoltage(DCDC, 0xB);
+	/* Enable 2.5 and 1.1V weak regulators */
+	PMU_2P5EnableWeakRegulator(PMU, true);
+	PMU_1P1EnableWeakRegulator(PMU, true);
+	/* Disable normal regulators */
+	PMU_2P5EnableOutput(PMU, false);
+	PMU_1P1EnableOutput(PMU, false);
+	/* Disable analog bandgap */
+	bandgap_set(false);
+}
+
+/* Undo the changes made by lpm_drop_voltage so clocks can be raised */
+static void lpm_raise_voltage(void)
+{
+	/* Enable analog bandgap */
+	bandgap_set(true);
+	/* Enable regulator LDOs */
+	PMU_2P5EnableOutput(PMU, true);
+	PMU_1P1EnableOutput(PMU, true);
+	/* Disable weak LDOs */
+	PMU_2P5EnableWeakRegulator(PMU, false);
+	PMU_1P1EnableWeakRegulator(PMU, false);
+	/* Change to 1.275V SOC voltage */
+	DCDC_AdjustRunTargetVoltage(DCDC, 0x13);
+	/* Move to the external RC oscillator */
+	CLOCK_InitExternalClk(0);
+	/* Switch clock source to external OSC. */
+	CLOCK_SwitchOsc(kCLOCK_XtalOsc);
+}
+
+
+/* Sets device into low power mode */
+__weak void pm_power_state_set(struct pm_state_info info)
+{
+	switch (info.state) {
+	case PM_STATE_RUNTIME_IDLE:
+		LOG_DBG("entering PM state runtime idle");
+		lpm_set_sleep_mode_config(kCLOCK_ModeWait);
+		lpm_enter_sleep_mode(kCLOCK_ModeWait);
+		break;
+	case PM_STATE_SUSPEND_TO_IDLE:
+		LOG_DBG("entering PM state suspend to idle");
```

Comment:
Not a exclusive problem of this patch, but logging at this point *may* be problematic. Device were already suspended here.